core/src/main/scala/shapeless/lazy.scala

```diff
@@ -46,48 +47,187 @@ object Lazy {
   def values[T <: HList](implicit lv: Lazy[Values[T]]): T = lv.value.values
 
   implicit def mkLazy[I]: Lazy[I] = macro LazyMacros.mkLazyImpl[I]
+
+  trait Cached[+T] extends Lazy[T] {
+    override def map[U](f: T => U): Cached[U] = Cached { f(value) }
+    def flatMap[U](f: T => Cached[U]): Cached[U] = Cached { f(value).value }
+  }
+
+  object Cached {
+    implicit def apply[T](t: => T): Cached[T] =
+      new Cached[T] {
+        val value = t
+      }
+
+    def unapply[T](lt: Cached[T]): Option[T] = Some(lt.value)
+
+    implicit def mkCached[I]: Cached[I] = macro LazyMacros.mkCachedLazyImpl[I]
+  }
+
 }
 
 object lazily {
   def apply[T](implicit lv: Lazy[T]): T = lv.value
 }
 
+trait Strict[+T] extends Serializable {
```

Comment:
Done in https://github.com/milessabin/shapeless/commit/a609a7e25ea561f494f76ecbd54e7c73a9b2f746